### PR TITLE
MetaDrive Vectorized Encoding Part III - Implement map + navigation perception

### PR DESCRIPTION
**Pull request by breakds**
# Motivation

This is the 3rd part of the effort to create the vectorized input for MetaDrive.

The environment need to provide observation for 

1. Map
2. Navigation
2. Ego history
2. (Other) Agent history

Map and navigation are both curves (i.e. represented as polylines). Therefore we can provide an unified perception module to generate both as feature vectors.

# Solution

Built upon the previous geometry utilities, this PR implements the `MapPolylinePerception`. The actual observation will employ `MapPolylinePerception` to generate the map + nav observation on each frame. The feature vectors is of shape `[B, F]`, where `B` is the number of polylines (capped by `polyline_limit`, so that it does not end up with too many polylines). A reasonable cap for 60m x 100m FOV is 128.

The feature for each polyline is a concatenation of

1. Features of all the segments that belongs to this polyline
2. A one-hot encoded category of the polyline's category (i.e. broken line, solid line, navigation, etc)

# Testing

Together with the other PRs (yet to come), had successful training with result on par with the BEV baseline.

### CODE REVIEW

File: alf/environments/metadrive/map_perception.py [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    scenario.
+
+    The essential method of MapPolylinePerception is observe(), which is called
+    upon every observation to generate the part of the observation from the map.
+
+    Under the hood, upon initialization we will create polylines for all of the
+    map objects. Whenever observe() is called, it will crop and rotate the
+    polylines (with batch vectorized operations) to produce the feature vectors
+    in a fast manner.
+
```
Comment (emailweixu) [inline alf/environments/metadrive/map_perception.py:44]: doc-string for arguments

File: alf/environments/metadrive/map_perception.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+        self._polylines = None
+
+    @property
+    def observation_spec(self):
+        return self._spec
+
+    def reset(self, road_network, navigation):
+        """Initialize by creating the polylines for all the map objects and navigation.
+
```
Comment (emailweixu) [inline alf/environments/metadrive/map_perception.py:100]: doc-string for arguments

File: alf/environments/metadrive/map_perception.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            self._polylines.category[i:end] = pl.category
+            i = end
+
+    def observe(self, position, heading) -> np.ndarray:
+        """Called upon every observation to get a rotated and cropped view of the map
+        objects and navigation. Returns the feature vector of the observation.
+
+        The position and heading of the observer car needs to be provided to
+        define the coordinate frame of the resulting features.
+
```
Comment (emailweixu) [inline alf/environments/metadrive/map_perception.py:166]: doc-string for arguments and return value

File: alf/environments/metadrive/map_perception.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+                polylines.append(
+                    Polyline.from_lane(lane, 0.0, category,
+                                       self._segment_resolution,
+                                       self._polyline_size))
+
+        # 3. Now polylines is a list of Polyline instances, where each Polyline
+        # instance corresponds to a batch of polylines. Here we flatten it so
+        # that it becomes a big batch of polylines within one single Polyline
+        # instance, which gets stored in self._polylines.
+        num_polylines = functools.reduce(lambda s, pl: s + pl.point.shape[0],
```
Comment (emailweixu) [inline alf/environments/metadrive/map_perception.py]: Perhaps the following is more readable:
```python
    num_polylines = sum(pl.point.shape[0] for pl in polylines)
```

File: alf/environments/metadrive/map_perception.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            point=np.zeros((num_polylines, self._polyline_size + 1, 2),
+                           dtype=np.float32),
+            category=np.zeros((num_polylines, ), dtype=np.int32))
+
+        i = 0
+        for pl in polylines:
+            end = i + pl.point.shape[0]
+            self._polylines.point[i:end] = pl.point
+            self._polylines.category[i:end] = pl.category
+            i = end
```
Comment (emailweixu) [inline alf/environments/metadrive/map_perception.py]: use `np.concatenate()`

**Review by emailweixu — COMMENTED** [review]
There are a lot of code in Part I-IV. I feel we should have some tests or visualization included.

**Comment by breakds** [comment]
> There are a lot of code in Part I-IV. I feel we should have some tests or visualization included.

Agree. I verified them by visualization, but it is a separate notebook. Integration of them and/or adding unit test may take some time but I'll work on them.

File: alf/environments/metadrive/map_perception.py (hunk 1 above)
Comment (breakds) [reply alf/environments/metadrive/map_perception.py:44]: Added.

File: alf/environments/metadrive/map_perception.py (hunk 2 above)
Comment (breakds) [reply alf/environments/metadrive/map_perception.py:100]: Added.

File: alf/environments/metadrive/map_perception.py (hunk 3 above)
Comment (breakds) [reply alf/environments/metadrive/map_perception.py:166]: Added.

File: alf/environments/metadrive/map_perception.py (hunk 4 above)
Comment (breakds) [reply alf/environments/metadrive/map_perception.py]: Alas, I did not realize python has builtin `sum`. Thanks!
Changed.

File: alf/environments/metadrive/map_perception.py (hunk 5 above)
Comment (breakds) [reply alf/environments/metadrive/map_perception.py]: Thanks! Updated.

**Comment by breakds** [comment]
> > There are a lot of code in Part I-IV. I feel we should have some tests or visualization included.
> 
> Agree. I verified them by visualization, but it is a separate notebook. Integration of them and/or adding unit test may take some time but I'll work on them.

After thinking about it, I think the most effective way is to add visualization and environment info support for those. I haven't pushed the final PRs and those visualization will be added.

**Review by Haichao-Zhang — DISMISSED** [review]

**Comment by breakds** [comment]
Similarly here, the last update is just a rebase on `pytorch`. Will need to be approved again. Thanks!

**Review by emailweixu — APPROVED** [review]

**Pull request merged**